cabal-install/Distribution/Client/VCS.hs

```diff
@@ -0,0 +1,559 @@
+import Distribution.Version
+         ( mkVersion )
+
+import Control.Monad
+         ( mapM_ )
+import qualified Data.Char as Char
+import qualified Data.Map  as Map
+import qualified Data.Set  as Set
+import Data.Set (Set)
+import Data.Ord
+         ( comparing )
+import System.FilePath
+         ( takeDirectory )
+import System.Directory
+         ( doesDirectoryExist )
+
+
+-- | A driver for a version control system, e.g. git, darcs etc.
+--
+data VCS program = VCS {
+       -- | The type of repository this driver is for.
+       vcsRepoType  :: RepoType,
+
+       -- | The vcs program itself.
+       -- This is used at type 'Program' and 'ConfiguredProgram'.
+       vcsProgram   :: program,
+
+       -- | The program invocation(s) to get\/clone a repository into a fresh
+       -- local directory.
+       vcsCloneRepo :: Verbosity
+                    -> ConfiguredProgram
+                    -> SourceRepo
+                    -> FilePath -- ^ Source URL
+                    -> FilePath -- ^ Destination directory
+                    -> [ProgramInvocation],
+
+       -- | The program invocation(s) to synchronise a whole set of /related/
+       -- repositories with corresponding local directories. Also returns the
+       -- files that the command depends on, for change monitoring.
+       vcsSyncRepos :: Verbosity
+                    -> ConfiguredProgram
+                    -> [(SourceRepo, FilePath)]
+                    -> IO [MonitorFilePath]
+     }
+
+
+-- | The set of all supported VCS drivers.
+--
+knownVCSs :: [VCS Program]
+knownVCSs = [ vcsBzr, vcsDarcs, vcsGit, vcsHg, vcsSvn ]
+
+
+-- ------------------------------------------------------------
+-- * Selecting repos and drivers
+-- ------------------------------------------------------------
+
+-- | Pick the 'SourceRepo' to use to get the package sources from.
+--
+-- Note that this does /not/ depend on what 'VCS' drivers we are able to
+-- successfully configure. It is based only on the 'SourceRepo's declared
+-- in the package, and optionally on a preferred 'RepoKind'.
+--
+selectPackageSourceRepo :: Maybe RepoKind
+                        -> [SourceRepo]
+                        -> Maybe SourceRepo
+selectPackageSourceRepo preferredRepoKind =
+    listToMaybe
+    -- Sort repositories by kind, from This to Head to Unknown. Repositories
+    -- with equivalent kinds are selected based on the order they appear in
+    -- the Cabal description file.
+  . sortBy (comparing thisFirst)
+    -- If the user has specified the repo kind, filter out the repositories
+    -- they're not interested in.
+  . filter (\repo -> maybe True (repoKind repo ==) preferredRepoKind)
+  where
+    thisFirst :: SourceRepo -> Int
+    thisFirst r = case repoKind r of
+        RepoThis -> 0
+        RepoHead -> case repoTag r of
+            -- If the type is 'head' but the author specified a tag, they
+            -- probably meant to create a 'this' repository but screwed up.
```

Comment:
Not entirely sure how ideal it is to help people write incorrect project configs.